src/overmap.cpp

```diff
@@ -4244,95 +4227,63 @@ void overmap::place_specials_pass( std::vector<std::pair<const overmap_special *
-        const std::string unplaced = enumerate_as_string( mandatory.begin(), mandatory.end(),
-        []( const std::pair<const overmap_special *, int> &elem ) {
-            return string_format( "%s (%d)", elem.first->id.c_str(), elem.second );
-        } );
-        dbg( D_WARNING ) << string_format( "couldn't place mandatory overmap specials during first pass (with city checks): %s.", unplaced.c_str() );
-
-        // Couldn't place mandatory specials, try without city checks
-        place_specials_pass( mandatory, sectors, false );
-        current_validity = overmap_valid::unlimited;
-    }
-
-    place_specials_pass( optional, sectors, true );
-
-    unplaced_mandatory_specials.clear();
-    if( !mandatory.empty() ) {
-        current_validity = overmap_valid::invalid;
-    }
-
-    const auto &unplaced = mandatory.empty() ? mandatory_unlimited : mandatory;
-    if( !unplaced.empty() ) {
-        const std::string unplaced_s = enumerate_as_string( mandatory.begin(), mandatory.end(),
-        []( const std::pair<const overmap_special *, int> &elem ) {
-            return string_format( "%s (%d)", elem.first->id.c_str(), elem.second );
-        } );
-        dbg( D_WARNING ) << string_format( "couldn't place mandatory overmap specials: %s.", unplaced_s.c_str() );
-        std::transform( unplaced.begin(), unplaced.end(), std::back_inserter( unplaced_mandatory_specials ),
-        []( const std::pair<const overmap_special *, int> &pr ) {
-            return pr.first;
-        } );
+    // First insure that all minimum instance counts are met.
+    place_specials_pass( enabled_specials, sectors, false );
+    if( std::any_of( enabled_specials.begin(), enabled_specials.end(),
+                     []( overmap_special_placement placement ) {
+                         return placement.instances_placed < placement.special_details->occurrences.min;
+                     } ) ) {
+        // Randomly select from among the nearest uninitialized overmap positions.
+        int previous_distance = 0;
+        std::vector<point> nearest_candidates;
+        for( point candidate_addr : closest_points_first( 10, pos() ) ) {
+            if( !overmap_buffer.has( candidate_addr.x, candidate_addr.y ) ) {
+                int current_distance = square_dist( pos().x, pos().y, candidate_addr.x, candidate_addr.y );
+                if( nearest_candidates.empty() || current_distance == previous_distance ) {
+                    nearest_candidates.push_back( candidate_addr );
+                    previous_distance = current_distance;
+                } else {
+                    break;
+                }
+            }
+        }
+        if( !nearest_candidates.empty() ) {
+            std::random_shuffle( nearest_candidates.begin(), nearest_candidates.end() );
+            point new_om_addr = nearest_candidates.front();
+            overmap_buffer.create_custom_overmap( new_om_addr.x, new_om_addr.y, enabled_specials );
+        }
```

Comment:
> If an object fails to generate within 2 overmap radius from the overmap where it was supposed to be, it must print a warning.

This is incorrect behavior in the rather mundane scenario of a player walking in a large circle and then exploring the center, at some point it will fail to generate additional overmaps due to all available locations being populated, but in reality there will already be sufficient specials in range.

> If it keeps generating overmaps indefinitely, there is no way of checking if the settings/specials are good enough to actually ever produce an overmap.

We can easily terminate recursion and warn on failure to place even a single mandatory special in an overmap, but that isn't a full solution.

> Also, it would be more "robust" if the generate_custom_overmap tried to cluster the new overmaps around the starting point instead of always going in one direction (up, I think).

It doesn't, reread overmap.cpp lines 4260-4283

None of your proposals recognize that overmap generation can and will fail occasionally due to exhaustion of nearby overmaps, and that this is not a bug.  If you don't get this, this discussion is utterly unproductive.  Also I'm far less concerned with theoretical bugs caused by misbehaving mods than I am by the status quo, which is that users are presented with a dialog asking them to retry mapgen indefinitely or accept a broken game world.

I'm inclined to adjust special definitions to include their terrain types and cut the generation range, but allow pre-existing specials to count toward the requisite totals, and warn on failure to either generate or find the required specials, but not in this pr.